### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
#purchase_data

## Player Count

* Display the total number of players


In [2]:
total_players = len(purchase_data.SN.unique())
#total_players

In [3]:
total_players_df =  pd.DataFrame({"Total Number of Players": total_players}, index =[0])
total_players_df

Total Number of Players
0                      576

# Some notes about the number of users

There are 780 observations, however there are only 576 total players, which means that there are players that bought more than one item.

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
Number_unique_items = len(purchase_data["Item Name"].unique())
#Number_unique_items

In [5]:
Average_purchase_price = purchase_data.Price.mean()
#Average_purchase_price

In [6]:
Total_number_purchase = len(purchase_data.Price)
#Total_number_purchase

In [7]:
Total_Revenue = purchase_data.Price.sum()
#Total_Revenue

In [8]:
purchasing_analysis_total = pd.DataFrame({"Number of Unique Items": Number_unique_items,
                                    "Average Purchase Price": Average_purchase_price,
                                          "Total Number of Purchases": Total_number_purchase,
                                          "Total Reveneu": Total_Revenue}, index = [0])
#purchasing_analysis_total

In [9]:
#Lets put formating in the money columns
purchasing_analysis_total["Average Purchase Price"] =  purchasing_analysis_total["Average Purchase Price"].map("${: .2f}".format)
purchasing_analysis_total["Total Reveneu"] =  purchasing_analysis_total["Total Reveneu"].map("${: .2f}".format)
purchasing_analysis_total


Number of Unique Items Average Purchase Price  Total Number of Purchases  \
0                     179                 $ 3.05                        780   

  Total Reveneu  
0     $ 2379.77

# Some notes about the items.

There are 179 purchasable items and 780 purchases which means that each item could be bought on average at least 3 times. How ever, it is important to do a better analysis of these numbers.

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [10]:
#Sort dataframe
prueba = purchase_data.sort_values('SN', ascending=True)
#prueba

In [11]:
#Drop duplicates sn
prueba_2 = prueba.drop_duplicates(subset ="SN", keep='first') 
  
# displaying data 
#prueba_2 

In [12]:
#gROUP BY GENDER
gender_gpby = prueba_2.groupby(["Gender"])
#gender_gpby

In [13]:
#Count how many male, female, or other make a purchase
Gender_group = gender_gpby.count()
#Gender_group

In [14]:
#Transform to a percentage the last count
Gender_group["Percentage"]= (Gender_group["Price"]/sum(Gender_group["Price"]))*100
#Gender_group

In [15]:
Gender_analysis = Gender_group.reset_index()
#Gender_analysis

In [16]:
#Remove undesired columns
Gender_analysis =  Gender_analysis.drop(labels=["Purchase ID","SN","Age","Item ID","Item Name"], axis=1)
#Gender_analysis

In [17]:
#Rename columns
Gender_analysis = Gender_analysis.rename(columns={"Price": "Count"})
#Gender_analysis                                                  

In [18]:
#Lets give formating to percentage
Gender_analysis["Percentage"] = Gender_analysis["Percentage"].map("%{: .2f}".format)
Gender_analysis

Gender  Count Percentage
0                 Female     81    % 14.06
1                   Male    484    % 84.03
2  Other / Non-Disclosed     11     % 1.91

# Some analysis about the section

It has to be noted that the data frame is mostly composed by Male, which could mean that the game is more attractive to this gender.


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [19]:
#Lets group the complete data by gender
new_group = purchase_data.groupby(['Gender'])
#new_group

In [20]:
#Lets count hou many purchases is done by Male, Female or other
purchase_count = new_group["Purchase ID"].count()
#purchase_count

In [21]:
#Lets get the average price per geneder
avg_purchase_price = new_group["Price"].mean()
#avg_purchase_price

In [22]:
#Lets get the total sum of purchases by gender
total_purchase_value = new_group['Price'].sum()
#total_purchase_value

In [23]:
#Lets get the total purchase per person and gender
purchase_total_person = total_purchase_value/gender_gpby['Price'].count()
#purchase_total_person

In [24]:
#Lets put all info in a df
purchasing_analysis_gender = pd.DataFrame({"Purchase count": purchase_count,
                                    "Average purchase price": avg_purchase_price,
                                          "Total purchase price": total_purchase_value,
                                          "Average Purchase Total per Person by Gender": purchase_total_person})
#purchasing_analysis_gender.head()

In [25]:
#Formating
purchasing_analysis_gender["Average purchase price"] = purchasing_analysis_gender["Average purchase price"].map("${: .2f}".format)
purchasing_analysis_gender["Total purchase price"] = purchasing_analysis_gender["Total purchase price"].map("${: .2f}".format)
purchasing_analysis_gender["Average Purchase Total per Person by Gender"] = purchasing_analysis_gender["Average Purchase Total per Person by Gender"].map("${: .2f}".format)
purchasing_analysis_gender

Purchase count Average purchase price  \
Gender                                                         
Female                            113                 $ 3.20   
Male                              652                 $ 3.02   
Other / Non-Disclosed              15                 $ 3.35   

                      Total purchase price  \
Gender                                       
Female                            $ 361.94   
Male                             $ 1967.64   
Other / Non-Disclosed              $ 50.19   

                      Average Purchase Total per Person by Gender  
Gender                                                             
Female                                                     $ 4.47  
Male                                                       $ 4.07  
Other / Non-Disclosed                                      $ 4.56

# Analysis of data 

Since we have more male players then we have more purchases from this gender. However, there the other two groups spent more money (more research required) in average.

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [26]:
#Lets create the bins
min_age=min(purchase_data["Age"])
max_age = max(purchase_data["Age"])
print(min_age, max_age)
bins =[5,9.99,14.99,19.99,24.99,29.99,34.99,39.99,45]
#print(bins)

7 45


In [27]:
#Lets create the groups
groups_name =["<10", "10-14","15-19", "20-24","25-29", "30-34", "35-39", "40+"]
#groups_name

In [28]:
#Apply the division in groups over non duplicated df
purchase_data["Age group"] = pd.cut(prueba_2["Age"], bins, labels=groups_name, include_lowest=True )
#purchase_data.head()

In [29]:
#Lets group by age group
data_group = purchase_data.groupby(["Age group"])
#data_group

In [30]:
#Lets count how many purchases are per age ide
Purchase_count_age = data_group.count()
Purchase_count_age = Purchase_count_age.drop(columns=["SN", "Age", "Gender", "Item ID", "Item Name", "Price"])
Purchase_count_age.rename(columns={"Purchase ID" : "Total Count"}, inplace=True)
#Purchase_count_age

In [31]:
#Get the precentage of player per age group
Purchase_count_age["Percentage of Players"] = (Purchase_count_age["Total Count"]/sum(Purchase_count_age["Total Count"]))*100
#Purchase_count_age

In [32]:
#Formating
Purchase_count_age["Percentage of Players"] = Purchase_count_age["Percentage of Players"].map("%{: .2f}".format)
Purchase_count_age

Total Count Percentage of Players
Age group                                   
<10                 17                % 2.95
10-14               22                % 3.82
15-19              107               % 18.58
20-24              258               % 44.79
25-29               77               % 13.37
30-34               52                % 9.03
35-39               31                % 5.38
40+                 12                % 2.08

# Some analysis about the data above

We already know that most of the buyers are Male. In total, they spent more money than the other groups. Another interesting fact is that almos half of the players have between 20-24 years old. The age group seems to have a normal distribution with the center in the age group 20-24.

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [33]:
purchase_data["Age group"] = pd.cut(purchase_data["Age"], bins, labels=groups_name, include_lowest=True )
#purchase_data.head()

In [34]:
#Group by age group the new partition (complete data)
data_group_2 = purchase_data.groupby(["Age group"])
#data_group_2

In [35]:
#Lets count number of purchase by age group
purchase_count_age = data_group_2["SN"].count()

#purchase_count_age

In [36]:
#Lets get the average purchase count by age group
average_purchase_c = data_group_2["Price"].mean()
#average_purchase_c

In [37]:
#Sum the total pruchase by age
total_purchase_c = data_group_2["Price"].sum()
#total_purchase_c

In [38]:
#Lets get the average purchase per person in each age group
average_p_total_person = total_purchase_c/ data_group['Price'].count()
#average_p_total_person

In [39]:
#Lets create a summary data frame
age_demographics = pd.DataFrame({"Purchase count" : purchase_count_age, 
                                "Average Purchase Price" : average_purchase_c,
                                "Total Purchase Value" : total_purchase_c, 
                                "Avg Total Purchase per Person":average_p_total_person })
#age_demographics

In [40]:
#Formating money columns
age_demographics["Average Purchase Price"] = age_demographics["Average Purchase Price"].map("$ {: .2f}".format)
age_demographics["Total Purchase Value"] = age_demographics["Total Purchase Value"].map("$ {: .2f}".format)
age_demographics["Avg Total Purchase per Person"] = age_demographics["Avg Total Purchase per Person"].map("$ {: .2f}".format)
age_demographics

Purchase count Average Purchase Price Total Purchase Value  \
Age group                                                               
<10                    23                $  3.35             $  77.13   
10-14                  28                $  2.96             $  82.78   
15-19                 136                $  3.04            $  412.89   
20-24                 365                $  3.05           $  1114.06   
25-29                 101                $  2.90            $  293.00   
30-34                  73                $  2.93            $  214.00   
35-39                  41                $  3.60            $  147.67   
40+                    13                $  2.94             $  38.24   

          Avg Total Purchase per Person  
Age group                                
<10                             $  4.54  
10-14                           $  3.76  
15-19                           $  3.86  
20-24                           $  4.32  
25-29                           $  3.81  
30-34                           $  4.12  
35-39                           $  4.76  
40+                             $  3.19

# Some highlights of data above

The 20 to 24 years group is the one whose total purchase is greater. But it seems that there is no clear difference in the average purchase per group, becasuse the values are very close.

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [41]:
# Lets group by SN which is a unique name of the player. We use the complete df
player_gp = purchase_data.groupby(["SN"])
#player_gp

In [42]:
#Lets count the number of purchases by player
purchase_count = player_gp["Price"].count()
#purchase_count

In [43]:
#Lets get the average purchase
purchase_avg = player_gp["Price"].mean()
#purchase_avg

In [44]:
#Lets get the total purchase per player
purchase_total = player_gp["Price"].sum()
#purchase_total

In [45]:
#Lets create the data frame
top_spenders = pd.DataFrame({"Purchase Count": purchase_count,
                            "Average Purchase Price": purchase_avg,
                            "Total Purchase Value" : purchase_total})
#top_spenders

In [46]:
#Lets sort by the highest total purchase value
top_spenders.sort_values('Total Purchase Value', ascending=False, inplace=True)
top_spenders


Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93                 5                3.792000                 18.96
Idastidru52               4                3.862500                 15.45
Chamjask73                3                4.610000                 13.83
Iral74                    4                3.405000                 13.62
Iskadarya95               3                4.366667                 13.10
...                     ...                     ...                   ...
Ililsasya43               1                1.020000                  1.02
Irilis75                  1                1.020000                  1.02
Aidai61                   1                1.010000                  1.01
Chanirra79                1                1.010000                  1.01
Alo38                     1                1.000000                  1.00

[576 rows x 3 columns]

In [47]:
#Format money columns
top_spenders["Average Purchase Price"] = top_spenders["Average Purchase Price"].map("$ {: .2f}".format)
top_spenders["Total Purchase Value"] = top_spenders["Total Purchase Value"].map("$ {: .2f}".format)
#top_spenders


In [48]:
#Kepp the top 5
top_spenders_5 = top_spenders.iloc[0:5]
top_spenders_5

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                $  3.79             $  18.96
Idastidru52               4                $  3.86             $  15.45
Chamjask73                3                $  4.61             $  13.83
Iral74                    4                $  3.40             $  13.62
Iskadarya95               3                $  4.37             $  13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [49]:
# Lets group by Item ID and Item Name which are a unique name for selled Item. We use the complete df
item_gp = purchase_data.groupby(["Item ID", "Item Name"])
#item_gp

In [50]:
#Lets get the count of how many itmes where bought by id
purchase_count = item_gp["Price"].count()
#purchase_count

In [51]:
#Lets get the item price
item_price = item_gp["Price"].mean()
#item_price

In [52]:
#Lets get the total purchase value
total_pv = item_gp["Price"].sum()
#total_pv

In [53]:
#Lets create the data frame
popular_item = pd.DataFrame({"Purchase Count": purchase_count,
                            "Item Price" : item_price,
                            "Total Purchase Value" : total_pv})
#popular_item

In [54]:
#Lest order the data frame in descending order to identify the most popular
popular_itemo = popular_item
popular_itemo.sort_values(['Purchase Count'], ascending = False, inplace=True)
#popular_itemo

In [55]:
#Formating columns of money
popular_itemo["Item Price"] = popular_itemo["Item Price"].map("${: .2f}".format)
popular_itemo["Total Purchase Value"] = popular_itemo["Total Purchase Value"].map("${: .2f}".format)
#popular_itemo

In [56]:
#Keep the top 5
popular_item_5 = popular_itemo.iloc[0:5]
popular_item_5

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                     $ 4.61   
178     Oathbreaker, Last Hope of the Breaking Storm     $ 4.23   
145     Fiery Glass Crusader                             $ 4.58   
132     Persuasion                                       $ 3.22   
108     Extraction, Quickblade Of Trembling Hands        $ 3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                              $ 59.99  
178     Oathbreaker, Last Hope of the Breaking Storm              $ 50.76  
145     Fiery Glass Crusader                                      $ 41.22  
132     Persuasion                                                $ 28.99  
108     Extraction, Quickblade Of Trembling Hands                 $ 31.77

# Analizyng above data

It seems that the most popular items have an averge price of 4 dollars. Perhaps, this is the price with which players feel comfortable to acquire items.

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [57]:
#Lets create the data frame
profitable_item= pd.DataFrame({"Purchase Count": purchase_count,
                            "Item Price" : item_price,
                            "Total Purchase Value" : total_pv})
#profitable_item

In [58]:
#Lets format columns again
profitable_item.sort_values('Total Purchase Value', ascending = False, inplace = True)
#profitable_item


In [59]:
#Formating columns of money
profitable_item["Item Price"] = profitable_item["Item Price"].map("${: .2f}".format)
profitable_item["Total Purchase Value"] = profitable_item["Total Purchase Value"].map("${: .2f}".format)
#profitable_item

In [60]:
#Keep the top 5
profitable_item_5 = profitable_item.iloc[0:5]
profitable_item_5

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                     $ 4.61   
178     Oathbreaker, Last Hope of the Breaking Storm     $ 4.23   
82      Nirvana                                          $ 4.90   
145     Fiery Glass Crusader                             $ 4.58   
103     Singed Scalpel                                   $ 4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                              $ 59.99  
178     Oathbreaker, Last Hope of the Breaking Storm              $ 50.76  
82      Nirvana                                                   $ 44.10  
145     Fiery Glass Crusader                                      $ 41.22  
103     Singed Scalpel                                            $ 34.80

# Analyzing above data

The trend in this data seems to confirm that 4 dollars in average is the money that players are willing to spent in each item.